typescript-for-haskellers/article.md

```diff
@@ -0,0 +1,726 @@
+Let's introduce more cases and how you can deal with them in TypeScript.
+
+```typescript
+type A = {
+  x: number;
+  y: number;
+};
+
+type ImmutableA = Readonly<A>; 
+const a: ImmutableA = { x: 1, y: 2 };
+a.x = 2; // error
+a.y = 1; // error
+
+
+const arr: ReadonlyArray<number> = [1, 2, 3];
+arr[0] = 4; // error
+
+
+interface A {
+    readonly [x: string]: number;
+};
+
+const a: A = { "x": 1, "y": 2 };
+a["x"] = 2; // error
+a["y"] = 1; // error
+```
+
+Also be careful passing object with readonly fields to function as it may cause unexpected behaviour.
+
+```typescript
+const a: { readonly x: number } = { x: 1 };
+
+const changeA = (arg: { x: number }): void => {
+  const argC = arg;
+  argC.x = 2;
+}
+
+changeA(a);
+console.log(a.x); // 2
+```
+
+Now you get more flexible way to make things readonly.
+Although now you can get immutable objects usually you don't need them as is.
+You want to make some actions on them and make new immutable objects.
+But it is not so easy.
+You will need to construct new objects by hands, you can also use `Writable` utility type, but it still is not really simple to copy and construct new objects.
+
+Let's introduce some libraries which may help you deal with this issue.
+The first one and the most powerful is [immer](https://immerjs.github.io/immer/).
+It is used for immutable state flow.
+Immutable data is not copied but shared in memory.
+It also has helpers for React.
+
+```typescript
+type A = Immutable<
+  {
+    text: string;
+    valid: boolean;
+  }[]
+>;
+
+const a: A = [
+  {
+    text: "valid text",
+    valid: true,
+  },
+  {
+    text: "invalid text",
+    valid: false,
+  },
+];
+
+const nextA = produce(a, (draft) => {
+  draft[1].valid = false;
+  draft.push({ text: "some text", valid: false });
+});
+```
+
+And finally there are three more interesting libraries for working with immutable structures which may be more familiar for Haskell devs: [partial.lenses](https://github.com/calmm-js/partial.lenses), [monocle-ts](https://github.com/gcanti/monocle-ts), and [optics-ts](https://github.com/akheron/optics-ts).
+These are optics libraries.
+Unfortunately the first one doesn't have type bindings.
+The second one support TypeScript and is a partial porting of Scala monocle.
+And the last one is the newest and TypeScript oriented optics library.
+
+### High Order functions and Currying
+
+A lot of languages support High Order functions and TypeScript is not an exception.
+You can use functions as an argument and also return them.
+
+```typescript
+const filterArray = (fn: (elem: string) => boolean, arr: string[]): string[] => {
+  const newArray: string[] = [];
+  for (let i = 0; i < arr.length; i++) {
+    if (fn(arr[i])) {
+      newArray.push(arr[i]);
+    }
+  }
+  return newArray; // or just simple return arr.filter(fn);
+};
+
+const f = (x: string): ((s: number) => boolean) => {
+  return (s: number) => {
+    return s.toString() === x;
+  };
+};
+
+f("5")(5);
```

Comment:
Here and further, could you also mention (e.g. in comments) what is returned/outputed by the call?